### DISCUSSION TITLE
Recommended `buildApprovedNamespaces` usage

### DISCUSSION THREAD

**Discussion by estrattonbailey** [WalletConnect/walletconnect-monorepo#2446] [Developer Support] [answered]
Hey thanks for the [new `buildApprovedNamespaces` util](https://docs.walletconnect.com/2.0/web/web3wallet/wallet-usage#-namespaces-builder-util), works great. I noticed, however, that it throws if certain aspects of the required namespace(s) aren't supported.

Is the recommended approach to catch thrown errors and then reject the session with an appropriate response? How can I determine what type the thrown error is? E.g. did it throw because an RPC method is unsupported, or because of a internal WC error. In these two example cases I'd want to handle the unhappy path differently, so it's important to be able to differentiate the two possible results.

Would be great to have an example of this pattern in the docs, as well as a note that this function will throw errors under certain conditions.

**Comment by ganchoradkov** [comment]
Hey @estrattonbailey, the error thrown already contains a message describing the issue e.g. not supported chains/methods/events, etc. 
By design, you should just show the error to the user but if you wish to handle the cases differently it's up to you.
The message is dynamically composed e.g. dapp requires chains `x,y,z` - wallet approves `x,y` vs dapp requires `x,y` - wallet approves `x` so you might be able to use regex or some other means to match the different cases.
You can find the possible messages here
https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/utils/src/validators.ts#L387

**Reply by estrattonbailey** [reply]
Yep yep, the error message is very helpful for showing to the user. Just to clarify, what would be helpful imo is differentiating between an _expected_ error and an _unexpected_ error. A common pattern you're probably experienced with is to subclass `Error` with like, `ValidationError`, so that the dev can use `isntanceof` to check which was thrown.

Perhaps in either unhappy case we want to reject the session. That's fine. Just calling out that it'd be helpful to have more information on this in the docs. Thanks!

**Reply by ganchoradkov** [reply]
Ah, I understand what you mean. Actually, we already have these although a bit different. Instead of a subclass, we have error codes that you can check here to differentiate between each case
https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/utils/src/errors.ts#L10

**Reply by ganchoradkov** [reply]
> Just calling out that it'd be helpful to have more information on this in the docs. Thanks!

Thanks for the suggestion, will add these to the docs asap 💯

**Reply by ganchoradkov** [reply, marked as answer]
https://docs.walletconnect.com/2.0/specs/clients/sign/error-codes
